activation/activation.go

```diff
@@ -496,133 +496,176 @@ func (b *Builder) BuildNIPostChallenge(ctx context.Context, nodeID types.NodeID)
 			zap.Time("waiting until", wait),
 		)
-		events.EmitPoetWaitRound(nodeID, current, publish, wait)
+		events.EmitPoetWaitRound(nodeID, currentEpochId, publishEpochId, wait)
 		select {
 		case <-ctx.Done():
 			return nil, ctx.Err()
 		case <-time.After(time.Until(wait)):
 		}
 	}
 
+	// 4. build new challenge
+	var challenge *types.NIPostChallenge
+
 	prevAtx, err = b.GetPrevAtx(nodeID)
 	switch {
 	case errors.Is(err, sql.ErrNotFound):
 		logger.Info("no previous ATX found, creating an initial nipost challenge")
-		post, err := nipost.GetPost(b.localDB, nodeID)
-		if err != nil {
-			return nil, fmt.Errorf("get initial post: %w", err)
-		}
-		logger.Info("verifying the initial post")
-		initialPost := &types.Post{
-			Nonce:   post.Nonce,
-			Indices: post.Indices,
-			Pow:     post.Pow,
-		}
-		err = b.validator.PostV2(ctx, nodeID, post.CommitmentATX, initialPost, shared.ZeroChallenge, post.NumUnits)
-		if err != nil {
-			logger.Error("initial POST is invalid", zap.Error(err))
-			if err := nipost.RemovePost(b.localDB, nodeID); err != nil {
-				logger.Fatal("failed to remove initial post", zap.Error(err))
-			}
-			return nil, fmt.Errorf("initial POST is invalid: %w", err)
-		}
-		posAtx, err := b.getPositioningAtx(ctx, nodeID, publish, nil)
+
+		challenge, err = b.buildInitialNIPostChallenge(ctx, logger, nodeID, publishEpochId)
 		if err != nil {
-			return nil, fmt.Errorf("failed to get positioning ATX: %w", err)
-		}
-		challenge = &types.NIPostChallenge{
-			PublishEpoch:   publish,
-			Sequence:       0,
-			PrevATXID:      types.EmptyATXID,
-			PositioningATX: posAtx,
-			CommitmentATX:  &post.CommitmentATX,
-			InitialPost: &types.Post{
-				Nonce:   post.Nonce,
-				Indices: post.Indices,
-				Pow:     post.Pow,
-			},
+			return nil, err
 		}
 	case err != nil:
 		return nil, fmt.Errorf("get last ATX: %w", err)
 	default:
 		// regular ATX challenge
-		posAtx, err := b.getPositioningAtx(ctx, nodeID, publish, prevAtx)
+		posAtx, err := b.getPositioningAtx(ctx, nodeID, publishEpochId, prevAtx)
 		if err != nil {
 			return nil, fmt.Errorf("failed to get positioning ATX: %w", err)
 		}
 		challenge = &types.NIPostChallenge{
-			PublishEpoch:   publish,
+			PublishEpoch:   publishEpochId,
 			Sequence:       prevAtx.Sequence + 1,
 			PrevATXID:      prevAtx.ID(),
 			PositioningATX: posAtx,
 		}
 	}
+
 	logger.Info("persisting the new NiPOST challenge", zap.Object("challenge", challenge))
 	if err := nipost.AddChallenge(b.localDB, nodeID, challenge); err != nil {
 		return nil, fmt.Errorf("add nipost challenge: %w", err)
 	}
 	return challenge, nil
 }
 
+func (b *Builder) getExistingChallenge(
+	logger *zap.Logger,
+	currentEpochId types.EpochID,
+	nodeID types.NodeID) (*types.NIPostChallenge, error) {
+	challenge, err := nipost.Challenge(b.localDB, nodeID)
+
+	switch {
+	case errors.Is(err, sql.ErrNotFound):
+		logger.Info("building new NiPOST challenge", zap.Uint32("current_epoch", currentEpochId.Uint32()))
```

Comment:
I think this log is misplaced now. How about moving it under `// 4. build new challenge`?